src/os/ObjectStore.h

```diff
@@ -766,6 +767,36 @@ class ObjectStore {
     const ghobject_t &oid  ///< [in] object
     ) = 0;
 
+  /**
+   * Iterate object map with user-provided callable
+   *
+   * Warning!  The callable is executed under lock on filestore
+   * operations in c.  Do not use filestore methods on c while
+   * when iterating.  (Filling in a transaction is no problem).
+   *
+   * @return error code, zero on success
+   */
+  struct omap_iter_seek_t {
+    std::string seek_position;
+    enum {
+      LOWER_BOUND,
+      UPPER_BOUND
+    } seek_type = LOWER_BOUND;
+    static omap_iter_seek_t min_lower_bound() { return {}; }
+  };
+  enum class omap_iter_ret_t {
+    STOP,
+    NEXT
+  };
+  virtual int omap_iterate(
```

Comment:
I like the concept of omap iteration as a black box.
It would be so much better if we do not expose RocksDB iterator behaviours to OSD.
Yes, i really propose to delete `OmapIterator` class altogether.

Reply:
I'm on board with this! `omap_iter_ret_t` already provides room for e.g. `PREV`; similarly `SEEK_TO_FIRST` could be added to `omap_iter_seek_t`.

Another interface-layer cleanup would be dropping (some of) the `std::string`-centric methods of `KeyValueDB::IteratorImpl` (then hopefully used only by providers of `ObjectStore`).

All of this as follow-ups with the reason being backportability – personally I'm fine even with short-living duplication of the code if it helps minimizing intrusiveness.

Reply:
https://github.com/ceph/ceph/pull/60890